ion-hash/src/type_qualifier.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+///! Implements "type qualifiers" (TQ) as per the [spec][spec].
+///!
+///! [spec]: https://amzn.github.io/ion-hash/docs/spec.html.
+use ion_rs::{
+    binary::IonTypeCode,
+    value::{AnyInt, Element},
+    IonType,
+};
+use num_bigint::Sign;
+
+pub(crate) struct TypeQualifier(u8);
+
+/// From the spec:
+///
+/// > TQ: is a type qualifier octet consisting of a four-bit type code T
+/// followed by a four-bit qualifier Q
+///
+/// To compute a TQ from a `T` and a `Q`, all we need to is a bitwise shift!
+const fn combine(ion_type_code: IonTypeCode, q: u8) -> TypeQualifier {
+    let t = ion_type_code.to_u8();
+    TypeQualifier(t << 4 | q)
+}
+
+impl TypeQualifier {
+    /// Computes a [`TypeQualifier`] from an [`Element`] according to the rules
+    /// laid out in the spec. In many cases, the `T` is determined by the Ion
+    /// binary type code.
+    ///
+    /// This code implements the TQ computation logically, but leans on `const
+    /// fn` to have the same runtime performance as if we wrote a jump table out
+    /// by hand.
```

Comment:
I don't think this is a jump table, but just a compile time generated lookup table, right?  I just want to make sure I am not misunderstanding the `const fn` above.

Reply:
yeah, let me update that